swarmrl/observables/col_graph.py

```diff
@@ -0,0 +1,243 @@
+        box_size=None,
+        record_memory=False,
+        particle_type: int = 0,
+    ):
+        """
+        Parameters
+        ----------
+        cutoff : float
+            Cutoff distance for the graph.
+        box_size : ndarray
+            Size of the box.
+        """
+        self.cutoff = cutoff
+        self.relation_angle = relation_angle
+        if box_size is None:
+            self.box_size = 1000 * np.array([1.0, 1.0, 1.0])
+        else:
+            self.box_size = box_size
+        self.eps = 10e-8
+
+        self.vangle = vmap(calc_signed_angle_between_directors, in_axes=(None, 0))
+        self.record_memory = record_memory
+        self.particle_type = particle_type
+
+        self.num_nodes = None
+        self.initialize(colloids)
+
+    def initialize(self, colloids: List[Colloid]):
+        self.num_nodes = len(colloids) - 1
+
+    def compute_initialization_input(self, colloids: List[Colloid]) -> GraphObservable:
+        """
+        A method that computes the input for the initialization of a graph model.
+        It is used to create the train-state of the graph model.
+        Instead of a graph with padded nodes for all colloids, it only
+        contains a single colloid. So to speak a unvectorized version of the
+        compute_observable method.
+
+        Parameters
+        ----------
+        colloids : List[Colloid]
+            List of colloids in the system.
+
+        Returns
+        -------
+        GraphObservable
+            A graph of a single colloid's observation.
+        """
+        system_graph = self.compute_observable(colloids)
+        # get the graph for the first colloid.
+        single_graph = GraphObservable(
+            nodes=system_graph.nodes[0],
+            edges=system_graph.edges[0],
+            destinations=system_graph.destinations[0],
+            receivers=system_graph.receivers[0],
+            senders=system_graph.senders[0],
+            globals_=system_graph.globals_[0],
+            n_node=system_graph.n_node[0],
+            n_edge=system_graph.n_edge[0],
+        )
+        return single_graph
+
+    def compute_observable(self, colloids: List[Colloid]) -> GraphObservable:
+        """
+        Builds a graph for each colloid in the system. In the graph, each node is a
+        representation of a colloid within the cutoff distance.
+        """
+        # normalize the positions by the box size.
+        positions = np.array([col.pos for col in colloids]) / self.box_size
+        types = np.array([col.type for col in colloids])
+        # compute the direction between all pais of colloids.
+        distances, part_part_vec = compute_dists_vecs(positions)
+
+        nodes_list = []
+        destinations_list = []
+        receivers_list = []
+        senders_list = []
+        n_nodes_list = []
+        n_edges_list = []
+        globals_list = []
+
+        for col in colloids:
+            if col.type == self.particle_type:
```

Comment:
I added some helper methods to collect these indices without the need for an if statement.